apps/test/unit/tutorialExplorer/tutorialSetTest.jsx

```diff
@@ -0,0 +1,215 @@
+import React from 'react';
+import {shallow} from 'enzyme';
+import {expect} from '../../util/configuredChai';
+import Tutorial from '@cdo/apps/tutorialExplorer/tutorial';
+import TutorialSet from '@cdo/apps/tutorialExplorer/tutorialSet';
+import TutorialDetail from '@cdo/apps/tutorialExplorer/tutorialDetail';
+import i18n from '@cdo/tutorialExplorer/locale';
+
+const TUTORIAL_1 = {
+  name: 'Tutorial 1',
+  code: 'tutorial-1',
+};
+
+const TUTORIAL_2 = {
+  name: 'Tutorial 2',
+  code: 'tutorial-2',
+};
+
+const DEFAULT_PROPS = {
+  tutorials: [],
+  filters: {},
+  localeEnglish: true,
+  disabledTutorials: [],
+};
+
+describe('TutorialSet', () => {
+  it('renders empty set of tutorials', () => {
+    const wrapper = shallow(
+      <TutorialSet {...DEFAULT_PROPS}/>
+    );
+    const instance = wrapper.instance();
+    expect(wrapper).to.containMatchingElement(
+      <div>
+        <TutorialDetail
+          showing={false}
+          item={null}
+          closeClicked={instance.tutorialDetailClosed}
+          localeEnglish={true}
+          disabledTutorial={false}
+          changeTutorial={instance.changeTutorial}
+        />
+        <div>
+          {i18n.tutorialSetNoTutorials()}
+        </div>
+      </div>
+    );
+  });
+
+  it('renders non-empty set of tutorials', () => {
+    const wrapper = shallow(
+      <TutorialSet
+        {...DEFAULT_PROPS}
+        tutorials={[TUTORIAL_1, TUTORIAL_2]}
+      />
+    );
+    const instance = wrapper.instance();
+    expect(wrapper).to.containMatchingElement(
+      <div>
+        <TutorialDetail
+          showing={false}
+          item={null}
+          closeClicked={instance.tutorialDetailClosed}
+          localeEnglish={true}
+          disabledTutorial={false}
+          changeTutorial={instance.changeTutorial}
+        />
+        <Tutorial
+          item={TUTORIAL_1}
+          filters={{}}
+          key={TUTORIAL_1.code}
+          tutorialClicked={wrapper.find(Tutorial).at(0).props().tutorialClicked}
+        />
+        <Tutorial
+          item={TUTORIAL_2}
+          filters={{}}
+          key={TUTORIAL_2.code}
+          tutorialClicked={wrapper.find(Tutorial).at(1).props().tutorialClicked}
+        />
+      </div>
+    );
+  });
+
+  it('shows tutorial details when tutorial is clicked', () => {
+    const wrapper = shallow(
+      <TutorialSet
+        {...DEFAULT_PROPS}
+        tutorials={[TUTORIAL_1, TUTORIAL_2]}
+      />
+    );
+
+    // Call the click callback directly.
+    wrapper.find(Tutorial).at(0).props().tutorialClicked();
```

Comment:
For my own curiosity, when should the click be called directly and when should a click be simulated?

Reply:
I'm not sure I'm right about this, but my thoughts going in were:

I think simulating events only works on a DOM component with a React-known event handler attached (in this case, the `<div>` inside `<Tutorial>` that has an `onClick` attribute).  Since we are shallow-rendering here we aren't actually creating that DOM component.  We can't simulate a click event directly on `Tutorial` because React doesn't know anything about the `tutorialClicked` attribute, we wrote that ourselves.  So the best we can do is assume `<Tutorial>` handles events and calls its callback correctly (or rather, test that in `tutorialTest.jsx`) and in this test just invoke the callback ourselves in the expected way.

There's probably a good argument that we should instead have used `mount()` here and rendered down to DOM components, in which case we could have simulated the click.